### PR TITLE
[lang] Added frontend type check for structs

### PR DESCRIPTION
**Pull request by listerily**
Issue: #8095 

### Summary

This PR added recursive type check for struct objects. It is implemented by magic method `__instancecheck__` on `StructType`. 

### ghstack

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #8150
* #8104
* #8149
* #8069
* __->__ #8092
* #8072
* #8099

### CODE REVIEW

**Comment by lin-hitonami** [comment]
Can we add some tests for the frontend type checks?

File: python/taichi/lang/struct.py [hunk 1]
```diff
@@ -649,7 +684,9 @@ def from_taichi_object(self, func_ret, ret_index=()):
                 d[name] = expr.Expr(_ti_core.make_get_element_expr(func_ret.ptr, ret_index + (index,)))
         d["__struct_methods"] = self.methods
 
-        return Struct(d)
+        struct = Struct(d)
+        struct._Struct_dtype = self.dtype
```
Comment (lin-hitonami) [inline python/taichi/lang/struct.py:688]: Should it be `_Struct__dtype`?

File: python/taichi/lang/struct.py [hunk 2]
```diff
@@ -670,7 +707,9 @@ def from_kernel_struct_ret(self, launch_ctx, ret_index=()):
                     raise TaichiRuntimeTypeError(f"Invalid return type on index={ret_index + (index, )}")
         d["__struct_methods"] = self.methods
 
-        return Struct(d)
+        struct = Struct(d)
+        struct._Struct_dtype = self.dtype
```
Comment (lin-hitonami) [inline python/taichi/lang/struct.py:711]: ditto

File: python/taichi/lang/struct.py [hunk 3]
```diff
@@ -707,7 +746,9 @@ def cast(self, struct):
                 else:
                     entries[k] = ops.cast(struct._Struct__entries[k], dtype)
         entries["__struct_methods"] = self.methods
-        return Struct(entries)
+        struct = Struct(entries)
+        struct._Struct_dtype = self.dtype
```
Comment (lin-hitonami) [inline python/taichi/lang/struct.py:750]: ditto

File: python/taichi/lang/struct.py [hunk 4]
```diff
@@ -719,11 +760,19 @@ def filled_with_scalar(self, value):
             else:
                 entries[k] = value
         entries["__struct_methods"] = self.methods
-        return Struct(entries)
+        struct = Struct(entries)
+        struct._Struct_dtype = self.dtype
```
Comment (lin-hitonami) [inline python/taichi/lang/struct.py:764]: ditto

File: python/taichi/lang/struct.py (hunk 1 above)
Comment (listerily) [reply python/taichi/lang/struct.py:688]: Sorry for the mistake.

**Review by lin-hitonami — APPROVED** [review]
LGTM

**Pull request merged**